### PR TITLE
DROOLS-4536 : Test scenarios fails with default values

### PR DESCRIPTION
**Pull request by Rikkola**
https://issues.jboss.org/browse/DROOLS-4536

@jomarko @manstis if you two could please review.
This presets the value that the textbox has by default for the model. I changed the style to MVP so it is testable.

### CODE REVIEW

File: drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/EditableTextBox.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        this.changed = changed;
+        this.view = view;
+        view.setPresenter(this);
+
+        view.setText(initialValue);
+        view.setTitle(fieldName);
+
+        if (initialValue == null || initialValue.isEmpty()) {
+            changed.valueChanged(view.getText());
+        }
```
Comment (Rikkola) [inline drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/EditableTextBox.java]: The actual fix is these 3 lines above this comment

**Comment by manstis** [comment]
@Rikkola in light of @jomarko comment please let me know when this PR is ready for review.

I assume you have to investigate and possible make additional changes.

**Comment by Rikkola** [comment]
@jomarko This works for me, but not sure if this bug is related:
https://issues.jboss.org/browse/DROOLS-4540

**Comment by Rikkola** [comment]
@jomarko If that bug was not the casue for it not working for you. I could use the test scenario you used.

File: drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/EditableTextBox.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import com.google.gwt.user.client.ui.Widget;
+import org.uberfire.client.callbacks.Callback;
+
+public class EditableTextBox
+        implements IsWidget {
+
+    private final EditableTextBoxView view;
+    private Callback<String> changed;
+
+    public EditableTextBox(final Callback<String> changed,
```
Comment (Rikkola) [inline drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/EditableTextBox.java]: @jomarko FYI I changed the deprecated listener to callback.

**Comment by Rikkola** [comment]
@manstis All good now. Ready for review.

File: drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/EditableTextBox.java (hunk 2 above)
Comment (jomarko) [reply drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/EditableTextBox.java]: @Rikkola Ok, I expect this change is safe.

**Comment by Rikkola** [comment]
So the `TextBox` you get from `TextBoxFactory.getTextBox(..)` has a preset value based on the text box type. When you create a new field check widget in the test scenario the `initialValue` has no value. So the UI and model are out of sync. The row that was linked syncs the model and UI when this happens. 

The complexity comes from me trying to move parts of this to MVP for testing purposes, while still leaving as much as it is so I don't break anything else.

Not sure if this explains it or causes more confusion ;)

**Comment by manstis** [comment]
@Rikkola Thank-you for the explanation.

I ran `master` code and debugged a little.

It appears that if you move the `ValueChangeHandler` registration in `VerifyFieldConstraintEditor` to before the value is set the model is correctly updated with the default/preset value. Moving [this](https://github.com/kiegroup/drools-wb/blob/master/drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/VerifyFieldConstraintEditor.java#L220) to the line after [this](https://github.com/kiegroup/drools-wb/blob/master/drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/VerifyFieldConstraintEditor.java#L216) also works successfully.

Please advise whether you prefer to handle the existing support for default values (in the presence of providing an invalid value) or whether you prefer to work around it, as this PR does... IMO moving the registration of the `ValueChangeHandler` is the _correct_ fix; but I'm an awkward fellow ;-)

**Comment by Rikkola** [comment]
Riiiight. That works. So setting the `TextBox` value to empty does not reset the `TextBox` value, but updates the model to what the `TextBox`default. Not sure why this was not the first thing I tried :)

This should be future proof too ( if this part of the code even has a future ) so that if the `TextBox` value someday can be set to empty, then it just updates to empty. I'll update the fix.

File: drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/test/java/org/drools/workbench/screens/testscenario/client/EditableTextBoxTest.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+        new EditableTextBox(valueChanged,
+                            view,
+                            "age",
+                            "11");
+
+        final InOrder o = inOrder(view, valueChanged);
+
+        o.verify(valueChanged).callback(anyString());
+        o.verify(view).setText("11");
```
Comment (Rikkola) [inline drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/test/java/org/drools/workbench/screens/testscenario/client/EditableTextBoxTest.java]: So now the actual fix is that these two happen in this order.

**Comment by Rikkola** [comment]
There is now quite a lot of complexity to just swap the order of two rows. Do we keep this for testing purposes or move out of this? I would prefer what the PR now contains, but open to suggestions.

File: drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/EditableTextBox.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+                           final EditableTextBoxView view,
+                           final String fieldName,
+                           final String initialValue) {
+        this.changed = changed;
+        this.view = view;
+        view.setPresenter(this);
+
+        changed.callback(view.getText());
```
Comment (manstis) [inline drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/EditableTextBox.java]: This should not be needed.

1. You add a `ValueChangeHandler` [here](https://github.com/kiegroup/drools-wb/pull/1231/files#diff-694563eaf1d207c6a6706ca0e0f9ab77R39).
2. When the value is set [here](https://github.com/kiegroup/drools-wb/pull/1231/files#diff-694563eaf1d207c6a6706ca0e0f9ab77R49), if it is invalid, the `ValueChangeHandler` attached above should be called.

So, provided, you call `view.setPresenter(this)` **before** `view.setText(initialValue)` (as you already do) then the existing mechanism to update the model with the default/valid value should execute.

Comment (Rikkola) [reply drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/EditableTextBox.java]: Right I actually meant to do this, but now this gives me another puzzle on how to test this in a way that makes sense. I'll figure something out.

**Comment by Rikkola** [comment]
@manstis Updated, let me know what you think.

@jomarko Not sure if you want to retry this or not.

**Comment by Rikkola** [comment]
@manstis always happy to see people review my changes with some effort. The current implementation is a lot better than what I proposed.

**Comment by Rikkola** [comment]
Jenkins please retest this.

**Comment by Rikkola** [comment]
@kiegroup/gatekeepers Can I get a merge please.

**Pull request merged**